### ISSUE TITLE
About the problems caused by Ultralytics YOLO version 8.3.111 and the Ultralytics version 8.3.63 provided by the author of the YOLOv12 paper

### ISSUE THREAD

**Issue by 202058304186** [ultralytics/ultralytics#20324]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

First of all, I have found out that the code provided by the author of the paper trains yolov12 with different results for each EPOCH, while the yolov12 officially provided by ultralytics can do the same every time. I have the following questions that I would like to have answered: 
1. Is there a difference between the turbo version and the standard version of the code except for the difference in the configuration function, which is inside the AAttn block?
2.  2. Why does the code provided by the authors show the use of USE_FLASH_ATTEN, while the official code from ultralytics doesn't show the use of it, and does this affect the performance of the two versions of the yolov12 performance performance?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @202058304186, thank you for reaching out and for your detailed questions about YOLO12 and related Ultralytics versions! 🚀 This is an automated response to help you get started—an Ultralytics engineer will follow up with you soon for a more in-depth discussion.

We suggest checking our [Docs](https://docs.ultralytics.com/) where you’ll find detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as explanations of core features and configuration options. Many common questions are addressed there.

If your inquiry relates to a 🐛 bug or implementation mismatch, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). This will help us better understand any differences you’re experiencing between versions or codebases.

For custom training or benchmarking ❓ questions, sharing detailed information—including configuration files, training logs, and dataset examples—will help us assist you faster. Also, refer to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for optimizing your experiments.

Join the Ultralytics community where it suits you best:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge sharing

## Upgrade

To ensure you’re running the latest YOLO12 and features, upgrade the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for your question and for helping improve Ultralytics!

**Comment by 202058304186** [comment]
> Do you train with Albumentations?
I suspect it has something to do with Flash-Attention, as I was able to reproduce each EPOCH when training other versions of YOLO with the paper author's code. The parameters of the training function used are as follows:
if __name__ == "__main__":
    # Load a COCO-pretrained YOLO12n model
    #model = YOLO(r"ultralytics\cfg\models\v12\Backbone\yolov12s-mobilenetv4.yaml")
    model = YOLO(r"ultralytics\cfg\models\v12\yolov12.yaml")
    # Train the model on the COCO8 example dataset for 100 epochs
    results = model.train(
        data=r"D:\Dataset\Blueberry\EI\data.yaml",
        cache=False,
        imgsz=640,
        epochs=300,
        batch=32,
        workers=4,
        device='0',
        optimizer='SGD', # using SGDs
        # patience=0, # close earlystop
        # resume=True, # 断点续训,YOLO初始化时选择last.pt2 
        deterministic = True,
        project=r'runs\v12\train\test',
        name="yolov12s-test"
        )

**Comment by Y-T-G** [comment]
I guess you need to manually pass `deterministic` option:

https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#how-to-use-flashattention

**Comment by 202058304186** [comment]
> 我猜你需要手动将 option 传递给 ：`deterministic``flash_attn_func`
> 
> https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#how-to-use-flashattention

Ok, thanks for your answer, I will try it. In addition to that, I would like to ask Ultralytics official code block.py doesn't explicitly use Flash-Attention inside, whereas the authors of the paper do, is this something that might have an impact on the model accuracy?

**Comment by glenn-jocher** [comment]
I'm glad you solved the reproducibility issue by setting the deterministic parameter in flash_attn_func to True! 

Regarding your question about FlashAttention impact: the Ultralytics YOLO12 implementation is designed to work without requiring FlashAttention by default, but it can optionally use it when available. This design choice helps with broader hardware compatibility, as FlashAttention requires specific NVIDIA GPUs (Turing, Ampere, Ada Lovelace, or Hopper).

Performance-wise, FlashAttention typically provides memory efficiency benefits rather than accuracy improvements - it reduces memory access overhead which can allow for faster training with larger batch sizes, but the model architecture and weights remain the same. Your experimental verification would be valuable to confirm this in your specific use case.

As a general note, we currently recommend YOLO11 over YOLO12 for most applications due to YOLO12's training instability, higher memory requirements, and slower inference speed (2-3x slower on CPU).

**Comment by 202058304186** [comment]
Thank you for your answers and suggestions, they are very useful to me.Hope everything goes well with your work.

**Comment by glenn-jocher** [comment]
You're very welcome! Thanks for the kind words and for being part of the Ultralytics community. Best wishes with your projects and experiments!
